src/overlay/Peer.cpp

```diff
@@ -1286,23 +1297,66 @@ Peer::recvGetTxSet(StellarMessage const& msg)
     if (auto txSet = mApp.getHerder().getTxSet(msg.txSetHash()))
     {
         StellarMessage newMsg;
-        newMsg.type(TX_SET);
-        txSet->toXDR(newMsg.txSet());
+        if (txSet->isGeneralizedTxSet())
+        {
+            if (mRemoteOverlayVersion <
+                Peer::FIRST_VERSION_SUPPORTING_GENERALIZED_TX_SET)
+            {
+                // The peer wouldn't be able to accept the generalized tx set,
+                // but it wouldn't be correct to say we don't have it. So we
+                // just let the request to timeout.
+                return;
+            }
+            newMsg.type(GENERALIZED_TX_SET);
+            txSet->toXDR(newMsg.generalizedTxSet());
+        }
+        else
+        {
+            newMsg.type(TX_SET);
+            txSet->toXDR(newMsg.txSet());
+        }
+
         auto newMsgPtr = std::make_shared<StellarMessage const>(newMsg);
         self->sendMessage(newMsgPtr);
     }
     else
     {
-        sendDontHave(TX_SET, msg.txSetHash());
+        // Technically we don't exactly know what is the kind of the tx set
+        // missing, however both TX_SET and GENERALIZED_TX_SET get the same
+        // treatment when missing, so it should be ok to maybe send the
+        // incorrect version during the upgrade.
+        auto messageType =
+            protocolVersionIsBefore(mApp.getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.ledgerVersion,
+                                    GENERALIZED_TX_SET_PROTOCOL_VERSION)
+                ? TX_SET
+                : GENERALIZED_TX_SET;
+        if (messageType == GENERALIZED_TX_SET &&
+            mRemoteOverlayVersion <
+                Peer::FIRST_VERSION_SUPPORTING_GENERALIZED_TX_SET)
+        {
+            return;
```

Comment:
mmm, any reason not to respond with TX_SET here? If the peer is on older overlay version, it doesn't understand generalized tx sets yet, so it has to be TX_SET

Reply:
the change still replies with generalized tx set as type. Maybe just remove line 1332 and let it timeout to be consistent with the behavior above

Reply:
It seems safer to reply with the old tx set for the sake of some weird catchup cases, doesn't it? I don't have a strong opinion on this, but unlike the case where we have tx set, giving a response here seems a bit more appropriate (as we don't say that we have something peer can't read).

Reply:
ok, that's fine, I think either behavior is appropriate. I see you updated this to be TX_SET, so I'm resolving this.